File: Notes/StrAndObj.swift
import Foundation

enum ObjType {
    case instance, str
}

struct Obj {
    let type: ObjType
}

typealias CharPtr = UnsafeMutableBufferPointer<UInt8>

extension CharPtr {
    init(capacity: Int) {
        let start = UnsafeMutablePointer<UInt8>.allocate(capacity: capacity)
        self.init(start: start, count: capacity)
    }
}

// "Subtype" of Obj
struct Str {
    let type: ObjType = .str
    let chars: CharPtr

    init(chars: CharPtr) {
        self.chars = chars
    }
}

typealias ObjPtr = UnsafePointer<Obj>
typealias StrPtr = UnsafePointer<Str>

extension StrPtr {
    init(chars: CharPtr) {
        let p = UnsafeMutablePointer<Str>.allocate(capacity: 1)
        p.initialize(to: Str(chars: chars))
        self.init(p)
    }
}

func unsafeUpcast<T>(_ value: UnsafePointer<T>) -> ObjPtr {
    // In C, `return (Obj *)(void *)value;`
    let raw = UnsafeRawPointer(value)
    return raw.bindMemory(to: Obj.self, capacity: 1)
}

extension ObjPtr {
    func asStr() -> StrPtr {
        assert(self.pointee.type == .str)
        // In C, `return (Str *)(void *)self;`
        let raw = UnsafeRawPointer(self)
        return raw.bindMemory(to: Str.self, capacity: 1)
    }
}

let bytes: [UInt8] = [0x48, 0x65, 0x6c, 0x6c, 0x6f, 0x2c, 0x20, 0x77, 0x6f, 0x72, 0x6c, 0x64, 0x21]
// buffer.len = /* find size of bytes */
// buffer.payload = malloc(sizeof(uint8_t) * buffer.len);
// memcpy(buffer.payload, bytes, buffer.len);
let chars = CharPtr(capacity: bytes.count)
_ = chars.initialize(from: bytes)

let sPtr = StrPtr(chars: chars)

print(sPtr.pointee)

let objPtr = unsafeUpcast(sPtr)
print(objPtr)
let str2 = objPtr.asStr()
print(str2)
print(str2.pointee)
print(str2.pointee.chars)
print(String(bytes: str2.pointee.chars, encoding: .utf8)!)

File: Notes/StrAndObj.c
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

typedef struct _buf {
    size_t len;
    uint8_t * payload;
} buf;

typedef struct _Obj {
    int kind;
} Obj;

#define AS_OBJ(o) ((Obj *)(o));
#define IS_STR(o) (((Obj *)(o))->kind == 1)

typedef struct _Str {
    Obj obj;
    buf chars;
} Str;

size_t arr_len(uint8_t * arr)
{
    size_t count = 0;
    while(*(arr + count) != 0x00) {
        count++;
    }
    
    return count;
}

Str * Str_make(buf buffer)
{
    Str * s = calloc(1, sizeof(Str));
    *s = (Str){ 1, buffer };
    return s;
}

void Str_print(Str * s)
{
    if (!(IS_STR(s))) {
        return;
    }
    
    printf("%s", s->chars.payload);
}


int main (int argc, char const *argv[])
{
    uint8_t bytes[] = { 0x48, 0x65, 0x6c, 0x6c, 0x6f, 0x2c, 0x20, 0x77, 0x6f, 0x72, 0x6c, 0x64, 0x21, 0x00 };
    buf buffer;
    buffer.len = arr_len(bytes);
    buffer.payload = calloc(buffer.len + 1, sizeof(uint8_t));
    memcpy(buffer.payload, bytes, buffer.len);
    
    Obj * o = AS_OBJ(Str_make(buffer));
    Str_print((Str *)o);
    
    return 0;
}
